main/src/mill/modules/Jvm.scala

```diff
@@ -257,20 +261,52 @@ object Jvm {
     manifest.write(manifestOut)
     manifestOut.close()
 
-    def isSignatureFile(mapping: String): Boolean =
-      Set("sf", "rsa", "dsa").exists(ext => mapping.toLowerCase.endsWith(s".$ext"))
-
-    for(v <- classpathIterator(inputPaths)){
-      val (file, mapping) = v
-      val p = zipFs.getPath(mapping)
-      if (p.getParent != null) Files.createDirectories(p.getParent)
-      if (!isSignatureFile(mapping)) {
-        val outputStream = newOutputStream(p)
-        IO.stream(file, outputStream)
-        outputStream.close()
+    val concatEntries: mutable.Map[String, List[AssemblyEntry]] = mutable.Map.empty
+    val seen: mutable.Set[String] = mutable.Set.empty
+
+    val rulesMap = assemblyRules.collect {
+      case r@Rule.Append(path) => path -> r
+      case r@Rule.Exclude(path) => path -> r
+    }.toMap
+    val appendPatterns = assemblyRules.collect {
+      case Rule.AppendPattern(pattern) => pattern.asPredicate().test(_)
+    }
+    val excludePatterns = assemblyRules.collect {
+      case Rule.ExcludePattern(pattern) => pattern.asPredicate().test(_)
+    }
+
+    for(entry <- classpathIterator(inputPaths)) {
+      val mapping = entry.mapping
+
+      def append() = concatEntries += mapping -> (entry :: concatEntries.getOrElse(mapping, Nil))
+
+      rulesMap.get(mapping) match {
+        case Some(_:Assembly.Rule.Append) =>
+          append()
+        case Some(_:Assembly.Rule.Exclude) =>
+          ctx.log.info(s"Excluding entry: ${mapping}")
+        case None =>
+          val path = zipFs.getPath(mapping)
+          if(!excludePatterns.exists(_(mapping))) {
+            if(appendPatterns.exists(_(mapping))) {
+              append()
+            } else if(!seen(mapping)) {
+              if(!Files.exists(path)) {
+                writeEntry(path, entry.is, append = false)
+              }
+              seen += mapping
+            }
+          }
       }
-      file.close()
     }
+
+    concatEntries.foreach { case (mapping, entries) =>
```

Comment:
Would it be possible to unify the `Some(_: Assembly.Rule.Append)` logic with the `None` logic, rather than having them go down a totally different code path? Apart from complexity, I think this would also re-arrange the order of files in the jar, which while harmless is unnecessary and makes debugging harder (e.g. when you `jar tf` two jars to compare contents)

Perhaps we could make the `None` case go through `concatEntries` as well, just over-writing a key rather than appending to it. If we then make `concatEntries` a `LinkedHashMap`, that should let us preserve the original ordering of keys

Reply:
The thing is - we process `concatEntries` outside the `for` comprehension that goes through all the classpath entries. If we process both "write once" and append outside `for` that means  we have to keep all the files in memory, before we write them. I guess we want to avoid this, right?

Reply:
@rockjam aren't we're only keeping the list of files in `concatEntries`, and not the file contents themselves, since each `AssemblyEntry` is just a thin wrapper that allows you to open an `InputStream` on-demand? As long as we're not keeping their byte contents, and not holding open all their file-handles at the same time, I think keeping the wrappers in memory should be acceptable. e.g. Ammonite only has `16710` file within it's assembly, and even an order of magnitude more 

Or we could just leave this as is; I guess a bit of reshuffling isn't the end of the world either.